### ISSUE TITLE
Try automatic machine translation

### ISSUE THREAD

**Issue by zeke** [electron/electron-i18n#6]
Crowdin can automatically apply machine translation to new content. This could be a big time-saver.

In order to use this, we'll need to add a machine translation API key to our Crowdin account from Google or Microsoft (or Yandex?)

<img width="617" alt="screen shot 2017-05-30 at 12 31 33 pm" src="https://cloud.githubusercontent.com/assets/2289/26603862/32d39ba2-453d-11e7-806b-efa54a5d1eaf.png">

https://support.crowdin.com/advanced-workflows/#accessing-the-workflows-feature

cc @alebourne

**Comment by alebourne** [comment]
HI @zeke 

I did some research on the translation memory and Crowdin. Both Google and Microsoft have neural machine translation and are fairly similar, however Microsoft is free up to 2 million characters in one month. I think that amount can cover the Electron use of machine translation, especially in the beginning when there are only a few languages.

**Comment by zeke** [comment]
I just set up [Microsoft Translator](https://www.microsoft.com/en-us/translator/getstarted.aspx) ,added the API key to our Crowdin project, and set up automatic machine translation for the Spanish locale.

@alebourne can you take a look at https://crowdin.com/project/electron/es-ES and tell me what you think of the translated Spanish?

Notice also that the [projects/electron](https://crowdin.com/project/electron) page has a nice callout saying that the machine translations need to be approved:

![screen shot 2017-05-30 at 9 20 20 pm](https://cloud.githubusercontent.com/assets/2289/26615735/155adbee-457e-11e7-90d0-e1d427ba9375.png)

## Cost

MS Translate is [free for the first 2 million characters](https://azure.microsoft.com/en-us/pricing/details/cognitive-services/translator-text-api/), and $10 for each million characters after that. They also have bundles for larger amounts.

This is this word count:

![screen shot 2017-05-30 at 9 16 19 pm](https://cloud.githubusercontent.com/assets/2289/26615741/1ff21f54-457e-11e7-968e-9b0731f52fb9.png)
 
(40,000 words) * (~5 letters per English word) = 200,000

So we can auto-translate roughly **ten locales** before we have to start paying for it. The next question I guess will be figuring out [which languages to prioritize for translation](https://github.com/electron/electron-i18n/issues/4).

**Comment by Toinane** [comment]
I suppose it most important to pre-translate languages which haven't translators yet or not much than other's languages.

**Comment by alebourne** [comment]
@zeke 
I think that @Toinane makes a good point about pre-translating some, but not all languages. Especially since it is possible for the person translating to use the machine translation on a string by string basis as needed.
With that said, machine translation is still not as accurate, so translations will need to be cleaned up. 
I will definitely have a look.

**Comment by alebourne** [comment]
@zeke, I checked 3 files and here are my observations:

- More than 95% of the translations in each file need to be "retranslated" (aka cleaned up). So doing the machine pre-translation is not really saving us from translating the text.
- The cleanup can be done in 2 ways. I can "vote down" the machine translation and propose a new one OR I can delete the machine translation and then type another one. In either case, another person needs to review and vote for the "new translation" that was added. Also; I'm not sure if everyone can "delete" translations, or if that is specific to my permissions in Crowdin.

Here are some additional thoughts:
- I think the best leverage/re-use will come from the translation memory. This will be available after the first Electron project is completed.

**Comment by zeke** [comment]
> doing the machine pre-translation is not really saving us

Thanks for this feedback. I wonder if we can easily undo the machine translation in bulk..

**Comment by Toinane** [comment]
I'm in the same case of @alebourne without permissions to delete pre-translation

**Comment by zeke** [comment]
Question to Crowdin support:

>  Is there a way to undo machine translations in bulk? We have done an experiment with machine translating Spanish and French, and would like to revert the machine translations without losing the human translations.

Crowdin response:

> Yes. Please select a language on the project page -> click the *Activity* tab -> click *Undo* across pre-translation action. Direct URLs:
> https://crowdin.com/project/electron/fr/activity
> https://crowdin.com/project/electron/es-ES/activity
> No worries about human translations, they will be preserved

**Comment by zeke** [comment]
<img width="1392" alt="screen shot 2017-06-02 at 8 30 50 am" src="https://cloud.githubusercontent.com/assets/2289/26732939/2efb3002-476e-11e7-84b8-bf85c30bc79c.png">

@Toinane I've removed the machine translations for French.

**Comment by funkyboy** [comment]
I went manually, basically editing the suggestions provided by crowdin. It's not that bad but it takes some time :) I finished just a file so far but I have no idea of the process.
Will a PR be sent to GH when the whole translation is finished (and I guess approved) on crowdin? Or you can approve/merge just one file (or a few) at a time?

**Comment by zeke** [comment]
> Will a PR be sent to GH when the whole translation is finished

Crowdin opens a single PR and automatically updates it every time a new translation is submitted on Crowdin to any file in any language. There's a way to configure Crowdin to require manager approval first, but that is disabled (by default). When a PR is merged, Crowdin opens a new one automatically.

Yesterday I set up some automated tests with Travis to make sure that translated files coming from Crowdin are properly named and formatted.

Here's an example of the most recent automated PR: https://github.com/electron/electron-i18n/pull/20

**Comment by zeke** [comment]
Gonna close this out.

TLDR:  Bulk machine translation seems to require just as much human work as translating from scratch, if not more. So for now we're disabling it.

**Comment by zeke** [comment]
The Azure trial account I created for Microsoft's Machine Translation service is expiring. No immediate plans to renew it.

<img width="673" alt="screen shot 2017-06-29 at 11 56 33 am" src="https://user-images.githubusercontent.com/2289/27705244-2a9e5f02-5cc2-11e7-82f4-914c48f7b3e9.png">

**Comment by vaartis** [comment]
Google translate actually works pretty good at least with English->Russian
